rust/kernel/src/miscdev.rs

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::error::{Error, KernelResult};
+use crate::file_operations::{FileOperations, FileOperationsVtable};
+use crate::{bindings, c_types, CStr};
+use core::marker::PhantomPinned;
+use core::pin::Pin;
+
+/// A registration of a misc device.
+pub struct Registration {
+    mdev: bindings::miscdevice,
+    registered: bool,
+    _pin: PhantomPinned,
+}
+
+impl Registration {
+    /// Initialises a new registration but does not register it yet. It is allowed to move.
+    pub fn new() -> Self {
+        Self {
+            mdev: bindings::miscdevice::default(),
+            registered: false,
+            _pin: PhantomPinned,
+        }
+    }
+
+    /// Attempts to actually register the misc device with the rest of the kernel. It must be
+    /// pinned because the memory block that represents the registration is self-referential. If a
+    /// minor is not given, the kernel allocates a new one if possible.
+    pub fn register<T: FileOperations>(
+        self: Pin<&mut Self>,
```

Comment:
Do you think we ought to adopt this pattern for chardev?

Reply:
Yes, absolutely. In fact, I think we need to do this for all self-referential data structures (e.g., mutexes). This is one big selling point for rust: we can express such things in the type system to avoid bugs.

Reply:
I opened issue #46 to track this.